### DISCUSSION TITLE
build for xtensa architecture (esp32, embedded) fails with confusing error

### DISCUSSION THREAD

**Discussion by crabdancing** [ipetkov/crane#759] [Q&A] [answered]
I'm working on a somewhat nontrivial project with multiple built targets, each of which have multiple architectures. [Here](https://gist.github.com/crabdancing/a9a4da47f547635ec6dab6112afd8b2d) is the build failure I'm getting.

Key point:

```
Caused by:
  process didn't exit successfully: `rustc - --crate-name ___ --print=file-names -C link-arg=-nostartfiles --target xtensa-esp32-none-elf --crate-type bin --crate-type rlib --crate-type dylib --crate-type cdylib --crate-type staticlib --crate-type proc-macro --print=sysroot --print=split-debuginfo --print=crate-name --print=cfg` (exit status: 1)
  --- stderr
  error: Error loading target specification: Could not find specification for target "xtensa-esp32-none-elf". Run `rustc --print target-list` for a list of built-in targets
```

If I run `rustc --print target-list`, I confirm that `xtensa-esp32-none-elf` _is_ in the build environment -- which is _also_ the toolchain package that crane is using. So I'm not sure why it's rejecting the target spec.

Note that it's using [this](https://github.com/crabdancing/esp-rs-nix) as the toolchain package, so it's quite possible that I messed something up or otherwise have introduced incompatibility with `crane` somehow. You can browse the repo [here](https://github.com/Ben-PH/SimpleFOC-rs/tree/52cdf0169bb2d372e8b176d634abd1402c448292), and see the flake in question [here](https://github.com/Ben-PH/SimpleFOC-rs/blob/52cdf0169bb2d372e8b176d634abd1402c448292/flake.nix).

Unfortunately, the flake is hard to meaningfully simplify because of the intrinsic complexity of targeting certain embedded devices.

To reproduce:

`nix build github:Ben-PH/SimpleFOC-rs?ref=52cdf0169bb2d372e8b176d634abd1402c448292#esp32_sfoc`

**Comment by ipetkov** [comment, marked as answer]
Hi @crabdancing thanks for the reproduction!

This looks like a configuration issue: although you've added `esp-rust` to the devshell, it isn't actually configured as the toolchain for the derivation. You'll likely need to fix one of the following:

* Set `craneLib = (inputs.crane.mkLib pkgs).overrideToolchain esp-rust;` [here](https://github.com/Ben-PH/SimpleFOC-rs/blob/52cdf0169bb2d372e8b176d634abd1402c448292/flake.nix#L39)
* Use `craneLibLLvmTools.buildPackage` [here](https://github.com/Ben-PH/SimpleFOC-rs/blob/52cdf0169bb2d372e8b176d634abd1402c448292/flake.nix#L110) (currently `craneLibLLvmTools` isn't used anywhere)

Hope this helps!

**Comment by crabdancing** [comment]
Thanks for the quick response @ipetkov! :) Sorry I missed that somewhat basic mistake.

After trying out the first one, I got this:

```
warning: Git tree '/home/ada/Projects/SimpleFOC-rs' is dirty
error: builder for '/nix/store/8a8rm2mrvj9w0qkcxr62549f7598khjg-esp32_sfoc-0.3.0.drv' failed with exit code 101;
       last 25 log lines:
       > Running phase: unpackPhase
       > unpacking source archive /nix/store/09qr3dj10s1a79y0ch9n9cjh9ymig5i9-source
       > source root is source
       > Running phase: patchPhase
       > code running
       > Executing configureCargoCommonVars
       > decompressing cargo artifacts from /nix/store/mrn53fzjr9785fbhrrpg2107bz0lk1m2-sfoc_rs-deps-0.3.0/target.tar.zst to target
       > Running phase: updateAutotoolsGnuConfigScriptsPhase
       > Running phase: configurePhase
       > will append /build/source/platforms/esp32_sfoc/.cargo-home/config.toml with contents of /nix/store/3r0znk0r9hzz62jxv9j1km09zkakmcan-vendor-cargo-deps/config.toml
       > default configurePhase, nothing to do
       > Running phase: buildPhase
       > ++ command cargo --version
       > cargo 1.79.0-nightly (ffa9cf99a 2024-06-03)
       > ++ command cargo build --release --message-format json-render-diagnostics -p esp32_sfoc
       > warning: profiles for the non root package will be ignored, specify profiles at the workspace root:
       > package:   /build/source/spinnies/esp32/original/Cargo.toml
       > workspace: /build/source/Cargo.toml
       > warning: profiles for the non root package will be ignored, specify profiles at the workspace root:
       > package:   /build/source/spinnies/esp32/s3/Cargo.toml
       > workspace: /build/source/Cargo.toml
       > error: no matching package named `compiler_builtins` found
       > location searched: registry `crates-io`
       > required by package `std v0.0.0 (/nix/store/d9r076qnij4h0b1dhm2hq1dchb59k1sp-esp-rs/lib/rustlib/src/rust/library/std)`
       >     ... which satisfies path dependency `std` (locked to 0.0.0) of package `sysroot v0.0.0 (/nix/store/d9r076qnij4h0b1dhm2hq1dchb59k1sp-esp-rs/lib/rustlib/src/rust/library/sysroot)`
       For full logs, run 'nix log /nix/store/8a8rm2mrvj9w0qkcxr62549f7598khjg-esp32_sfoc-0.3.0.drv'.
```

This also happens when using `craneLibLLvmTools` (see commit `#63bec0fddb9491654176f0fdbbc2b276911ca704`).

**Reply by ipetkov** [reply]
This is probably because the target does not come with a precompiled stdlib. Take a look at the [build-std example](https://github.com/ipetkov/crane/blob/604637106e420ad99907cae401e13ab6b452e7d9/examples/build-std/flake.nix) on how to achieve that
